qbft/round_change.go

```diff
@@ -120,66 +119,81 @@ func hasReceivedProposalJustificationForLeadingRound(
-		// If justifiedRoundChangeMsg has prepare justification chose prepared value
-		valueToPropose := instanceStartValue
-		if msg.Message.RoundChangePrepared() {
-			valueToPropose = signedRoundChange.FullData
-		}
-
 		roundChangeJustification, _ := msg.Message.GetRoundChangeJustifications() // no need to check error, checked on isValidRoundChange
-		if isProposalJustificationForLeadingRound(
+		if valueToPropose, err := isProposalJustificationForLeadingRound(
 			state,
 			config,
 			msg,
 			roundChanges,
 			roundChangeJustification,
-			valueToPropose,
+			cdFetcher,
 			valCheck,
 			signedRoundChange.Message.Round,
-		) == nil {
+		); err == nil {
 			// not returning error, no need to
 			return msg, valueToPropose, nil
 		}
 	}
 	return nil, nil, nil
 }
 
-// isProposalJustificationForLeadingRound - returns nil if we have a quorum of round change msgs and highest justified value for leading round
+// isProposalJustificationForLeadingRound - returns proposal value if we have a quorum of round change msgs and
+// highest justified value for leading round
 func isProposalJustificationForLeadingRound(
 	state *State,
 	config IConfig,
 	roundChangeMsg *SignedMessage,
 	roundChanges []*SignedMessage,
 	roundChangeJustifications []*SignedMessage,
-	value []byte,
+	cdFetcher *types.DataFetcher,
 	valCheck ProposedValueCheckF,
 	newRound Round,
-) error {
+) ([]byte, error) {
+	if proposer(state, config, roundChangeMsg.Message.Round) != state.Share.OperatorID {
+		return nil, errors.New("not proposer")
+	}
+
+	valueToPropose, err := chooseValueToPropose(roundChangeMsg, cdFetcher)
+	if err != nil {
+		return nil, err
+	}
+
 	if err := isReceivedProposalJustification(
 		state,
 		config,
 		roundChanges,
 		roundChangeJustifications,
 		roundChangeMsg.Message.Round,
-		value,
+		valueToPropose,
 		valCheck); err != nil {
-		return err
-	}
-
-	if proposer(state, config, roundChangeMsg.Message.Round) != state.Share.OperatorID {
-		return errors.New("not proposer")
+		return nil, err
 	}
 
 	currentRoundProposal := state.ProposalAcceptedForCurrentRound == nil && state.Round == newRound
 	futureRoundProposal := newRound > state.Round
 
 	if !currentRoundProposal && !futureRoundProposal {
-		return errors.New("proposal round mismatch")
+		return nil, errors.New("proposal round mismatch")
 	}
 
-	return nil
+	return valueToPropose, nil
+}
+
+// chooseValueToPropose
+// If justifiedRoundChangeMsg has no prepare justification choose state value
+// If justifiedRoundChangeMsg has prepare justification choose prepared value
+func chooseValueToPropose(roundChangeMsg *SignedMessage, cdFetcher *types.DataFetcher) ([]byte, error) {
+	var valueToPropose []byte
+	if roundChangeMsg.Message.RoundChangePrepared() {
+		valueToPropose = roundChangeMsg.FullData
+	} else {
+		cd, err := cdFetcher.GetConsensusData()
+		if err != nil {
+			return nil, err
+		}
+		valueToPropose = cd
+	}
+	return valueToPropose, nil
 }
```

Comment:
I believe this function can be called more than once during the Round-Change phase. If so, we would also wait for the Beacon Node more than once. Is there a way to save the data, e.g. storing it in the `i.StartValue`, so it can be returned faster?

Reply:
The idea of `cdFetcher` is that implementation can do it in a way where it calls it asynchronously and even caches values. But this is an implementation detail.

@olegshmuelov